posts/2024-11-25-cloudant-with-open-liberty.adoc

```diff
@@ -0,0 +1,170 @@
+
+== A CDI producer for Cloudant
+With a CDI producer, you can easily provide a Cloudant client to your application by which you can inject the client into various parts of the application in a type-safe and flexible way, while leveraging the benefits of dependency injection (such as lifecycle management and configuration).This example demonstrates how to create a CDI producer to inject a Cloudant client:
+[source, java]
+----
+@ApplicationScoped
+public class CloudantProducer {
+
+    @Produces
+    public Cloudant createCloudant() {
+        return new Cloudant(“cloudant”, new BasicAuthenticator.Builder().build());
+    }
+}
+----
+
+Here is an example of using the CDI producer to inject a Cloudant client in a JAX-RS application. 
+[source, java]
+----
+@Inject
+Cloudant client;
+@POST
+@Path("/add")
+@Consumes(MediaType.APPLICATION_JSON)
+public void add(CrewMember crewMember) {
+        Document newCrewMember = new Document();			 
+        newCrewMember.put("Name",crewMember.getName());
+        newCrewMember.put("Rank",crewMember.getRank());
+        newCrewMember.put("CrewID",crewMember.getCrewID());
+
+        PostDocumentOptions createDocumentOptions =
+                    new PostDocumentOptions.Builder()
+                        .db(dbname)
+                        .document(newCrewMember)
+                        .build();
+           DocumentResult createDocumentResponse = client
+                    	.postDocument(createDocumentOptions)
+                        .execute()
+                   	    .getResult();
+}
+----
+In the above code:
+
+* `@Inject` is used to inject the Cloudant client instance provided by the CDI producer.
+* The `postDocument` method demonstrates how to create and post a document in the cloudant database using the injected CloudantClient.
+
+== Enhancing the CDI producer
+One of the advantages of using a CDI producer is that it can be tailored to your needs. For improved security,the `createCloudant` method can be enhanced to include authentication with a user name and password:
+
+[source,java]
+----
+@Produces
+public Cloudant createCloudant() {
+     String password = PasswordUtil.passwordDecode(encodedPassword);
+     BasicAuthenticator authenticator = new BasicAuthenticator.Builder()
+                .username(username)
+                .password(password)
+                .build();
+
+    return new Cloudant("cloudant", authenticator);
+}
+----
+This requires `passwordUtilities-1.0`, available as a Maven dependency:
+[source,xml]
+----
+<dependency>
+    <groupId>io.openliberty.features</groupId>
+    <artifactId>passwordUtilities-1.0</artifactId>
+    <version>18.0.0.4</version>
```

Comment:
This uses password util now instead:
<dependency>
    <groupId>com.ibm.websphere.appserver.api</groupId>
    <artifactId>com.ibm.websphere.appserver.api.passwordUtil</artifactId>
    <version>1.0.95</version>
</dependency>